optimum/neuron/modeling_seq2seq.py

```diff
@@ -438,9 +517,15 @@ def generate(
             axis=1,
         )
 
-        # copy the new cache state to the decoder
-        for state, tensor in zip(self.decoder.model.parameters(), past_key_values):
-            state.copy_(tensor)
+        if not self.tp_degree > 1:
+            # copy the new cache state to the decoder
+            for state, tensor in zip(self.decoder.model.parameters(), past_key_values):
+                state.copy_(tensor)
+        else:
+            # Encoder returns cache as device tensors, we assign them to decoder's cache to avoid the copy.
+            # The KV cache always use pre-allocated memory, no host-device communication overhead.
+            for decoder_tp, encoder_tp in zip(self.decoder.model.models, self.encoder.model.models):
```

Comment:
decoder.model is the ScriptModule returned by parallel_model_trace, right ? Does it contains several sub-models corresponding to each rank ? Maybe that deserves a little explanation because I don't think there is any way to figure it out what these 'models' are just by reading the code (or I missed something).